### ISSUE TITLE
drain daemonsets

### ISSUE THREAD

**Issue by kfox1111** [kubernetes/kubernetes#75482]
**What would you like to be added**:
a flag to drain to support draining daemonset managed pods too.

**Why is this needed**:
I needed to blow out container runtime images completely to fix an issue. daemonsets tried to keep pods running.

**Comment by kfox1111** [comment]
Taking a guess here...
@kubernetes/sig-node

**Comment by nikopen** [comment]
/milestone v1.16
/sig apps
/kind bug
/priority important-soon

I'd say it's a bug as `drain` is not able to hold its promise of `--ignore-daemonsets=false` which is even the default value.

Prioritizing this for the next release, ideally

**Comment by paivagustavo** [comment]
Hey, I would like to give this a try. @nikopen do you think a newcomer do kubernetes could fix this issue? If yes, could you give me some directions?

**Comment by paivagustavo** [comment]
I was going through the docs and found [this statement](https://kubernetes.io/docs/concepts/architecture/nodes/):
> Note: Pods created by a DaemonSet controller bypass the Kubernetes scheduler and do not respect the unschedulable attribute on a node. This assumes that daemons belong on the machine even if it is being drained of applications while it prepares for a reboot.

Is this really a bug or is it the intended behavior?

**Comment by nikopen** [comment]
@paivagustavo I think the implementation is not a problem (likely trivial), the problem lies mostly in regards to decision-making from the responsible SIG. Could be viewed as both a bug and intented, but there *should* be a way to manually force this behavior.

**Comment by kfox1111** [comment]
Yeah. Generally, you want the network stack (kube-proxy,flannel,etc) to continue to function while draining the node of pods, so you don't want those to get deleted too early. But you do want a way to eventually completely clean all pods out. Typically I've seen upgrading docker/containerd/etc loose track of pods because of this issue, as kubelet still thinks they are on the node, but the engine no longer finds them. I hit it again today. So we do need a solution to it.

Maybe an additional kind of drain that taints the node as going completely off line and deletes all remaining pods? This would prevent the daemonsets from relaunching pods on them.

**Comment by xmudrii** [comment]
@kfox1111 Hello! I'm bug triage lead for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th (about 1.5 weeks from now), which means that there should be a PR ready (and merged) until then.

Do we still target this issue to be fixed for 1.16?

**Comment by kfox1111** [comment]
I hope so, but I don't think anyone is actively working on it at the moment. So probably will miss 1.16.

**Comment by ttousai** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on Nov. 18. Is this issue still intended for 1.17?

**Comment by kfox1111** [comment]
This still periodically affects me. Need a fix.

**Comment by ttousai** [comment]
***Correction:*** Code freeze is on Thursday, November 14.

Sorry for the mix up.

**Comment by derekwaynecarr** [comment]
Are you able to address this by performing a power cycle?  Typically, we do not recommend updating the host without rebooting it.

**Comment by kfox1111** [comment]
No. Sometimes upgrading the container engine may cause it to loose track of containers unsafely but persistently (storage). The instructions say, to clean off 'all' containers off of the system before upgrading the container runtime. Daemonsets are ignoring this and not actually draining. (Sometimes useful). There needs to be an easy way to completely remove all running kubernetes managed containers off of a host. Like, a flag on the host or something to say, kill all the rest and don't let it restart automatically.

**Comment by wilmardo** [comment]
We ran into this today, this issue is the result from a design decision at containerd:

> KillMode handles when containerd is being shut down. By default, systemd will look in its named cgroup and kill every process that it knows about for the service. This is not what we want. As ops, we want to be able to upgrade containerd and allow existing containers to keep running without interruption. Setting KillMode to process ensures that systemd only kills the containerd daemon and not any child processes such as the shims and containers.

Source: https://github.com/containerd/containerd/blob/master/docs/ops.md#systemd

So when containerd is upgraded it will abandon its child processes by design. So all the ignored daemonsets still running on the drained node are abandoned but Kubernetes will not pick them back up after uncordoning. This is seems the issue that @kfox1111 describes by:

> Typically I've seen upgrading docker/containerd/etc loose track of pods because of this issue, as kubelet still thinks they are on the node, but the engine no longer finds them

This is resolved by rebooting node as @derekwaynecarr suggests since this kill the abandoned child processes.

If needed I can provide some examples. In short this is reproducible by:

* Run a cluster on containerd (1.2.6 for example)
* Drain a node with --ignore-daemonsets
* Upgrade containerd to 1.2.10 on this node
* Run `ps faux` and see the abandoned child processes
* Uncordon the node and see Kubernetes not being able to resolve the state
* Restart the node
* Watch all pods go to running again

**Comment by kfox1111** [comment]
Containerd has it for sure. Other runtimes might as well. The need to drain a node completely of workload is a reasonable one on its own regardless of which runtime is involved though.

I've seen kubelet continuously complain in the logs about pods that disappeared after an upgrade of containerd as well, so just rebooting doesn't entirely fix the issue.

**Comment by gianarb** [comment]
Hello @kubernetes/sig-node-bugs    !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by gianarb** [comment]
Hello @kubernetes/sig-node-bugs @kubernetes/sig-apps-bugs  !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by kfox1111** [comment]
Still a bug. No one is working on it. Should probably be bumped to 1.19.

**Comment by gianarb** [comment]
Ok @kfox1111 I am moving it now.   Thanks

/milestone 1.19

**Comment by bai** [comment]
👋 Hello from Bug Triage team! Wanted to follow up on this issue with a friendly reminder that the code freeze for 1.19 is starting June 25th (about 4 weeks from now).

As this issue is tagged for 1.19, is it still planned for this release?

**Comment by kfox1111** [comment]
Still a bug. No one is working on it.

**Comment by hamishforbes** [comment]
Are there any workarounds for this issue in the meantime?
I have a slightly different use case. My k8s nodes are EC2 spot instances and I would like to *gracefully* terminate as many pods as possible when the spot instance is due to be terminated.
The [aws-node-termination-handler](https://github.com/aws/aws-node-termination-handler) automatically drains the node which works great for normal workloads but leaves daemonsets.

Specifically I want my Consul agent daemonset pod to exit so that it will leave the Consul cluster and not be left hanging as a failed node. 

For me an extra label or annotation that indicates a daemonset shouldn't have the default tolerations and therefore would be evicted, or some way to override those default tolerations would be fine.

Similar issue https://github.com/kubernetes/autoscaler/issues/1578

**Comment by cgwalters** [comment]
We're hitting this in OpenShift too, see https://bugzilla.redhat.com/show_bug.cgi?id=1884053 - basically we're exposing dns via a damonset + service, and if we don't delete the pod the endpoint can appear incorrectly active.

I think the simplest implementation of this should be that things like the aws termination handler (and the MCO) send SIGTERM to kubelet; that should in turn involve sending SIGTERM to any outstanding pods including daemonsets.  And when kubelet gets that it should delete the daemonset pod from the apiserver and avoid scheduling any new pods or so.

Handling it this way would require us to rework some things in the MCO because the actual post-drain update+reboot happens via the daemonset pod, but we can probably rework it to be a systemd unit since it's mostly operating on the host anyways.

**Comment by kfox1111** [comment]
Does this issue affect the 1.21 upgrade where it wants you to drain out all pods before switching from cgroupfs to systemd? Its recommended to drain the nodes before switching, but if you --ignore-daemonsets then they are left running.

**Comment by pacoxu** [comment]
> Typically, we do not recommend updating the host without rebooting it.

I prefer to fix it as a documentation issue: add some reminder to reboot after `kubectl drain`.

What if daemonsets tolerate all taints? It's really odd to invert a new definition to make `draining node` refuse all pods(even tolerate all taints) from scheduling to the node.

**Comment by kfox1111** [comment]
rebooting doesn't delete the pods though. so if the container runtime upgrade causes some differences in the filesystem that need to be corrected by actually fully draining all the containers off of the host, --ingnore-daemonsets won't do it?

**Comment by kfox1111** [comment]
A taint that can not be ignored to force the rest of the pods off of the host may be part of the solution.

**Comment by adisky** [comment]
Subscribe (Take a look at it later)

**Comment by adisky** [comment]
@liyanzhao The issue we were working on were kind of similar, right?

**Comment by kfox1111** [comment]
Still an issue. :/
/remove-lifecycle stale

**Comment by kfox1111** [comment]
Still an issue in 2022. :/
/remove-lifecycle stale

**Comment by furkatgofurov7** [comment]
Hi! Is there any update (or plans to add support) on this issue?

**Comment by furkatgofurov7** [comment]
@kfox1111 hi, how can we move forward with this issue, this has been here for a long time, any ideas? Is it a problem of no one willing to take this up or it needs a confirmation from sig node?
Also, @nikopen hi, is there anything I can do to move forward with this issue, thanks!

**Comment by kfox1111** [comment]
Pretty sure its not on anyone's radar anymore. Someone should bring it back up to sig-node maybe?

**Comment by furkatgofurov7** [comment]
> Pretty sure its not on anyone's radar anymore. Someone should bring it back up to sig-node maybe?

Got it, sorry I am not aware of the process, what exactly does one have to do to take this up to sig-node/'s attention?
/sig node

**Comment by kfox1111** [comment]
> Got it, sorry I am not aware of the process, what exactly does one have to do to take this up to sig-node/'s attention? /sig node

Sorry. I don't really know.

Guessing it may need to be brought up at one of the meetings?

**Comment by kfox1111** [comment]
still an issue. :/

**Comment by SergeyKanzhelev** [comment]
/priority backlog

to reflect the reality

**Comment by atiratree** [comment]
This feature might be eventually supported with Declarative Node Maintenance: https://github.com/kubernetes/enhancements/pull/4213

**Comment by SergeyKanzhelev** [comment]
/remove-lifecycle rotten

this still a feature request
